src/sentry_plugins/amazon_sqs/plugin.py

```diff
@@ -118,57 +133,76 @@ def forward_event(self, event, payload):
                 # if content based de-duplication is not enabled, we need to provide a
                 # MessageDeduplicationId
-                message["MessageDeduplicationId"] = uuid4().hex
+                send_message_args["MessageDeduplicationId"] = uuid4().hex
+            return client.send_message(**send_message_args)
+
+        # wrap S3 put_object and and SQS send message in one try/except
+        s3_bucket = self.get_option("s3_bucket", event.project)
+        try:
+            # if we have an S3 bucket, upload to S3
+            if s3_bucket:
+                # we want something like 2020-08-29 so we can store it by the date
+                date = event.datetime.strftime("%Y-%m-%d")
+                key = "{}/{}/{}".format(event.project.slug, date, event.event_id)
+                logger.info("sentry_plugins.amazon_sqs.s3_put_object", extra=logging_params)
+                s3_put_object(Bucket=s3_bucket, Body=json.dumps(payload), Key=key)
+
+                url = u"https://{}.s3-{}.amazonaws.com/{}".format(s3_bucket, region, key)
+                # just include the s3Url and the event ID in the payload
+                payload = {"s3Url": url, "eventID": event.event_id}
+
+            message = json.dumps(payload)
+
+            if len(message) > 256 * 1024:
+                logger.info("sentry_plugins.amazon_sqs.skip_oversized", extra=logging_params)
+                return False
+
             logger.info("sentry_plugins.amazon_sqs.send_message", extra=logging_params)
-            client.send_message(**message)
+            sqs_send_message(message)
         except ClientError as e:
             if six.text_type(e).startswith(
                 "An error occurred (InvalidClientTokenId)"
             ) or six.text_type(e).startswith("An error occurred (AccessDenied)"):
                 # If there's an issue with the user's token then we can't do
                 # anything to recover. Just log and continue.
-                metrics_name = "sentry_plugins.amazon_sqs.access_token_invalid"
-                logger.info(
-                    metrics_name, extra=logging_params,
-                )
-
-                metrics.incr(
-                    metrics_name, tags=metric_tags,
-                )
+                log_and_increment("sentry_plugins.amazon_sqs.access_token_invalid")
                 return False
             elif six.text_type(e).endswith("must contain the parameter MessageGroupId."):
-                metrics_name = "sentry_plugins.amazon_sqs.missing_message_group_id"
-                logger.info(
-                    metrics_name, extra=logging_params,
-                )
-                metrics.incr(
-                    metrics_name, tags=metric_tags,
-                )
+                log_and_increment("sentry_plugins.amazon_sqs.missing_message_group_id")
+                return False
+            elif six.text_type(e).startswith("An error occurred (NoSuchBucket)"):
```

Comment:
I think you should have separate try/catch blocks for S3 and SQS so that S3 errors don't block the queue.

Reply:
@mgaeta Well, if the S3 upload fails we shouldn't try uploading to SQS since we have no S3 reference to upload. I grouped them together to avoid having two different catch statements for the expired token.

Reply:
👍